io/jvm/src/main/scala/fs2/io/ioplatform.scala

```diff
@@ -144,4 +146,25 @@ private[fs2] trait ioplatform extends iojvmnative {
     }
   }
 
+  private lazy val vtExecutor: Option[ExecutionContext] = {
+    val javaVersion: Int =
+      System.getProperty("java.version").stripPrefix("1.").takeWhile(_.isDigit).toInt
+
+    // From JVM 21 on we can use virtual threads
+    if (javaVersion >= 21) {
+      val virtualThreadExecutor = classOf[Executors]
+        .getDeclaredMethod("newVirtualThreadPerTaskExecutor")
+        .invoke(null)
+        .asInstanceOf[ExecutorService]
+
+      ExecutionContext.fromExecutor(virtualThreadExecutor).some
+    } else {
+      None
+    }
+
+  }
+
+  def evalOnVirtualThreadIfAvailable[F[_]: Async, A](fa: F[A]): F[A] =
+    vtExecutor.fold(fa)(ec => fa.evalOn(ec))
```

Comment:
AFAIK accessing a `lazy val` is costly for the first initialization but not worse than a regular if check afterwards.

Reply:
> AFAIK accessing a `lazy val` is costly for the first initialization

That's correct.

> not worse than a regular if check afterwards.

Well, not quite. Because `lazy val`s are synchronized for thread-safe initialization, reading one is equivalent to reading a `@volatile val`, which can be more expensive than a non-volatile. But I don't think we need to worry too much about it, because working with processes already involves lots of synchronization and memory barriers.

---

If we really cared, instead of using `lazy val` we could use a `var` with our own non-thread-safe initialization scheme. In this case, we don't care if we create the VT executor more than once during initialization, it's not a big deal.